### DISCUSSION TITLE
Improve Memory Management

### DISCUSSION THREAD

**Discussion by zeuchste** [nebulastream/nebulastream#42] [Ideas]
### Description

The current memory management has many problems. One is that we can deadlock the system, this we should handle in the current milestone. 

The second one is that their can be many optimizations and fairness considerations in the memory management which are a subject of the next milestone.

## Deadlock

We currently have no deadlock detection/handling in NES. Sources are bound by a fix buffer pool. However, worker have a variable pool which means if their pool is exhausted, they will pull new buffers from the global buffer pool. This is not regulated and an eager worker thread can exhaust all buffers and the system stalls. Good examples where this will easily happen is a nested loop join. Also we do not detect deadlock or resolve it by for example kill one query.

There are two possible solutions

1. We throw an error if some components want to allocate a buffer and the pool is empty and thus kill the query
2. We bound the processing buffers pools also thus all components like pipelines or handler have a fix buffer pool too and will stall if all buffers are exhausted. This could create another deadlock then

I would favor solution 1

# Later considerations

## Fairness

Currently all queries/worker have the same right to pull all buffers this is maybe not a good thing as there are maybe more important queries.

## Per query configuration

Currently all queries have the same number of source buffers and processing buffer. However, it would make for example more sense to give stateless queries that require only a few buffers in the processing less buffers. Currently they get as many as a stateful query. As we distribute this on startup of the query, to the local buffer pool, the buffers are blocked and cannot be used by others

### Acceptance Criteria

We avoid deadlocks in the system and provide fairness among all queries. in addition we make sure that no memory is leaked

**Comment by keyseven123** [comment]
For now, I would also vote for solution 1. This is not by any stretch a good and an optimal algorithm but it should be easy to implement and we can then extent it later down the line. It also makes sure that we can never run into any deadlock, so it solves the biggest problem about our current memory management.
